"could not prepare statement" ("Connection is closed") when reusing a session for another transaction after a rollback

## Describe the bug

This sequence of actions causes ORM to reuse a connection that was closed implicitly by Agroal, ultimately throwing an exception:

* Open session manually
* Open JTA transaction manually
* Do stuff in ORM => ORM opens a transaction
* Roll back the transaction => ORM should have closed the connection, but it didn't, so Agroal closes the transaction without telling it.
* Start another transaction
* Do stuff in ORM => ORM thinks the connection is still open, so it reuses it, resulting in an exception.

This happens in particular when mass indexing in Hibernate Search, because Hibernate Search rolls back transactions instead of committing them, since it only reads data.

### Expected behavior

No exception. ORM should close connections on rollback, at least with the default connection handling mode.

### Actual behavior

Mass indexing fails after the first batch, with errors like this:

```
2021-04-13 08:53:01,407 ERROR [org.hib.sea.eng.rep.imp.LogFailureHandler] (Hibernate Search - Mass indexing - FruitEntity - Entity loading - 8) HSEARCH000058: Exception occurred javax.persistence.PersistenceException: org.hibernate.exception.GenericJDBCException: could not prepare statement
Failing operation:
Loading and extracting entity data for entity 'FruitEntity' during mass indexing
: javax.persistence.PersistenceException: org.hibernate.exception.GenericJDBCException: could not prepare statement
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1602)
	at org.hibernate.query.Query.getResultList(Query.java:165)
	at org.hibernate.query.criteria.internal.compile.CriteriaQueryTypeQueryAdapter.getResultList(CriteriaQueryTypeQueryAdapter.java:76)
	at org.hibernate.search.mapper.orm.massindexing.impl.IdentifierConsumerDocumentProducer.loadAndIndexList(IdentifierConsumerDocumentProducer.java:154)
	at org.hibernate.search.mapper.orm.massindexing.impl.IdentifierConsumerDocumentProducer.loadAndIndexAllFromQueue(IdentifierConsumerDocumentProducer.java:123)
	at org.hibernate.search.mapper.orm.massindexing.impl.IdentifierConsumerDocumentProducer.run(IdentifierConsumerDocumentProducer.java:103)
	at org.hibernate.search.util.common.impl.CancellableExecutionCompletableFuture$CompletingRunnable.run(CancellableExecutionCompletableFuture.java:49)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.hibernate.exception.GenericJDBCException: could not prepare statement
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:47)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:186)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:151)
	at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:2103)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2040)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2018)
	at org.hibernate.loader.Loader.doQuery(Loader.java:948)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:349)
	at org.hibernate.loader.Loader.doList(Loader.java:2849)
	at org.hibernate.loader.Loader.doList(Loader.java:2831)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2663)
	at org.hibernate.loader.Loader.list(Loader.java:2658)
	at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:506)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:400)
	at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:219)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1414)
	at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1625)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1593)
	... 12 more
Caused by: java.sql.SQLException: Connection is closed
	at io.agroal.pool.wrapper.ConnectionWrapper.lambda$static$0(ConnectionWrapper.java:51)
	at com.sun.proxy.$Proxy92.prepareStatement(Unknown Source)
	at io.agroal.pool.wrapper.ConnectionWrapper.prepareStatement(ConnectionWrapper.java:616)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:149)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:176)
	... 28 more
```

## To Reproduce

Originally [reported by @suchwerk](https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Hibernate.20Search.20-.20Massindexer): thanks to him!

Reproducer (in Quarkus): https://github.com/suchwerk/quarkus-hibernate-search-massindexer

You'll need to start an Elasticsearch container, then run the tests:

```
docker run -it --rm=true --name es-7.10-it -p 9200:9200 -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.0
./mvnw clean verify
```

### Configuration

N/A

### Screenshots

N/A

## Environment (please complete the following information):

### Quarkus version or git rev

1.13.1.Final

Probably applies to 1.13.x only, since we switched the connection handling mode in 1.13.0.Final.

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -93,8 +93,8 @@
         <commons-lang3.version>3.12.0</commons-lang3.version>
         <commons-codec.version>1.15</commons-codec.version>
         <classmate.version>1.5.1</classmate.version>
-        <hibernate-orm.version>5.4.30.Final</hibernate-orm.version>
-        <hibernate-reactive.version>1.0.0.CR3</hibernate-reactive.version>
+        <hibernate-orm.version>5.4.31.Final</hibernate-orm.version>
+        <hibernate-reactive.version>1.0.0.CR4</hibernate-reactive.version>
         <hibernate-validator.version>6.2.0.Final</hibernate-validator.version>
         <hibernate-search.version>6.0.3.Final</hibernate-search.version>
         <narayana.version>5.11.1.Final</narayana.version>

--- a/extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/ReactiveSessionProducer.java
+++ b/extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/ReactiveSessionProducer.java
@@ -34,13 +34,24 @@ public class ReactiveSessionProducer {
 
     public void disposeStageSession(@Disposes Stage.Session reactiveSession) {
         if (reactiveSession != null) {
-            reactiveSession.close();
+            //We're ignoring the returned CompletionStage!
+            //This could certainly done better but it should be effective enough to
+            //eventually close the session: the connection pool will order the operations
+            //as it won't be able to return connections to other consumers when it's saturated,
+            //ensuring close operations are handled.
+            reactiveSession.close().toCompletableFuture().join();
         }
     }
 
     public void disposeMutinySession(@Disposes Mutiny.Session reactiveSession) {
         if (reactiveSession != null) {
-            reactiveSession.close();
+            //We're ignoring the returned CompletionStage!
+            //This could certainly done better but it should be effective enough to
+            //eventually close the session: the connection pool will order the operations
+            //as it won't be able to return connections to other consumers when it's saturated,
+            //ensuring close operations are handled.
+            //N.B. for the Mutiny API we need to make sure to subscribe
+            reactiveSession.close().subscribe().asCompletionStage().join();
         }
     }
 }

--- a/extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/ThreadLocalPool.java
+++ b/extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/ThreadLocalPool.java
@@ -3,6 +3,7 @@ package io.quarkus.reactive.datasource.runtime;
 import java.lang.ref.WeakReference;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.CompletableFuture;
 
 import io.vertx.core.AsyncResult;
 import io.vertx.core.Future;
@@ -121,28 +122,22 @@ public abstract class ThreadLocalPool<PoolType extends Pool> implements Pool {
 
     @Override
     public void close(Handler<AsyncResult<Void>> handler) {
-        synchronized (allConnections) {
-            this.closed = true;
-            for (PoolAndThread pair : allConnections) {
-                pair.close();
-            }
-            allConnections.clear();
-            threadLocal.remove();
-        }
-        handler.handle(Future.succeededFuture());
+        handler.handle(close());
     }
 
     @Override
     public Future<Void> close() {
         synchronized (allConnections) {
             this.closed = true;
+            ArrayList<CompletableFuture<Void>> tasks = new ArrayList<>(allConnections.size());
             for (PoolAndThread pair : allConnections) {
-                pair.close();
+                tasks.add(pair.close().toCompletionStage().toCompletableFuture());
             }
+            final CompletableFuture<Void> combinedCloseTask = CompletableFuture.allOf(tasks.toArray(new CompletableFuture[0]));
             allConnections.clear();
             threadLocal.remove();
+            return Future.fromCompletionStage(combinedCloseTask);
         }
-        return Future.succeededFuture();
     }
 
     //Useful for testing mostly
@@ -195,9 +190,11 @@ public abstract class ThreadLocalPool<PoolType extends Pool> implements Pool {
 
         /**
          * Closes the connection
+         * 
+         * @return
          */
-        void close() {
-            pool.close();
+        Future<Void> close() {
+            return pool.close();
         }
 
         /**

--- a/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/basicproxy/AbstractEntity.java
+++ b/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/basicproxy/AbstractEntity.java
@@ -1,0 +1,43 @@
+package io.quarkus.it.jpa.h2.basicproxy;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+import javax.persistence.Column;
+import javax.persistence.DiscriminatorColumn;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "ENTITY")
+@DiscriminatorColumn(name = "TYPE")
+public abstract class AbstractEntity implements Serializable {
+
+    public AbstractEntity() {
+        // nop
+    }
+
+    @Id
+    @Column(name = "TYPE", insertable = false, updatable = false)
+    public String type;
+
+    @Id
+    @Column(name = "ID")
+    public String id;
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        AbstractEntity that = (AbstractEntity) o;
+        return Objects.equals(type, that.type) && Objects.equals(id, that.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(type, id);
+    }
+}

--- a/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/basicproxy/BasicProxyTestEndpoint.java
+++ b/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/basicproxy/BasicProxyTestEndpoint.java
@@ -1,0 +1,41 @@
+package io.quarkus.it.jpa.h2.basicproxy;
+
+import java.io.IOException;
+import java.util.List;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.transaction.Transactional;
+
+import io.quarkus.runtime.StartupEvent;
+
+@ApplicationScoped
+@WebServlet(urlPatterns = "/jpa-h2/testbasicproxy")
+public class BasicProxyTestEndpoint extends HttpServlet {
+
+    @Inject
+    EntityManager entityManager;
+
+    @Transactional
+    public void setup(@Observes StartupEvent startupEvent) {
+        ConcreteEntity entity = new ConcreteEntity();
+        entity.id = "1";
+        entity.type = "Concrete";
+        entityManager.persist(entity);
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        final List list = entityManager.createQuery("from ConcreteEntity").getResultList();
+        if (list.size() != 1) {
+            throw new RuntimeException("Expected 1 result, got " + list.size());
+        }
+        resp.getWriter().write("OK");
+    }
+}

--- a/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/basicproxy/ConcreteEntity.java
+++ b/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/basicproxy/ConcreteEntity.java
@@ -1,0 +1,14 @@
+package io.quarkus.it.jpa.h2.basicproxy;
+
+import javax.persistence.DiscriminatorValue;
+import javax.persistence.Entity;
+
+@Entity
+@DiscriminatorValue("Concrete")
+public class ConcreteEntity extends AbstractEntity {
+
+    public ConcreteEntity() {
+        super();
+    }
+
+}

--- a/integration-tests/jpa-h2/src/test/java/io/quarkus/it/jpa/h2/JPAFunctionalityTest.java
+++ b/integration-tests/jpa-h2/src/test/java/io/quarkus/it/jpa/h2/JPAFunctionalityTest.java
@@ -25,4 +25,9 @@ public class JPAFunctionalityTest {
         RestAssured.when().get("/jpa-h2/testproxy").then().body(is("OK"));
     }
 
+    @Test
+    public void testHibernateEnhancedBasicProxies() throws Exception {
+        RestAssured.when().get("/jpa-h2/testbasicproxy").then().body(is("OK"));
+    }
+
 }

